### PR TITLE
feat(instrumentation): add Anthropic Go SDK instrumentation (non-streaming)

### PR DESCRIPTION
**Pull request by pradhyum6144**
Adds compile-time instrumentation for github.com/anthropics/anthropic-sdk-go, producing GenAI client spans for the non-streaming Messages API. This is the first phase of #679 and follows the pattern established by the openai-go instrumentation.

The rule hooks anthropic.NewClient and injects an HTTP middleware via option.WithMiddleware, gated by the ANTHROPIC instrumentation key. The middleware classifies POST /v1/messages calls (the path suffix keeps count_tokens and batches out of scope), starts a client span, and suppresses the underlying net/http client span so there is no duplicate.

Attributes follow the gen_ai semantic conventions: request model, max_tokens, temperature, top_p, and top_k (top_k is Anthropic specific, with no openai equivalent), provider name, response id and model, finish reasons, and token usage. Anthropic returns a single stop_reason rather than per-choice finish_reason values, so it is wrapped into a one element slice to keep the shared RequireGenAIClientSemconv helper reusable. The API has no total_tokens field, so it is computed from input plus output. Prompt-cache usage (cache_read and cache_creation input tokens) is recorded when present.

Per the design discussion in #679, this covers only the non-streaming path. The streaming reader is left as a documented stub and will land in a follow-up once the accumulation approach is confirmed. Content capture is deferred to a later phase.

The rule is pinned to a v1.0.0 lower bound so the single module covers a broad version range.

Tested with unit tests covering the hook, operation classification, provider mapping, request parsing, and the full span shape including the cache-token cases. An integration test app and mock-server test mirror the openai-go setup and run under the existing integration CI.

Part of #679

### CODE REVIEW

**Review by NameHaibinZhang — COMMENTED** [review]
## Code Review Summary

Thanks for this PR! The implementation closely follows the OpenAI instrumentation pattern and covers the non-streaming Messages API well. The unit tests are thorough for the main paths.

### Findings

**Must fix before merge:**
- `total_tokens` calculation is incorrect when prompt caching is used — Anthropic's `input_tokens` excludes cache tokens (unlike OpenAI)
- Streaming requests emit incomplete spans missing all `gen_ai.response.*` and `gen_ai.usage.*` attributes

**Should fix:**
- Request body is partially consumed on `ReadAll` error, passing a truncated body to the SDK
- Missing nil guard on `resp.Body` in `handleNonStreamingResponse`
- `streamingReader` edge cases (close-before-read, double-close, empty stream) are untested

**Consider:**
- `getProviderName` uses map iteration with non-deterministic order
- Integration test doesn't verify Anthropic-specific cache attributes end-to-end
- Consider adding a `demo/app/anthropic/` for consistency with OpenAI

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+	span.SetAttributes(
+		semconv.GenAIResponseID(resp.ID),
+		semconv.GenAIResponseModel(resp.Model),
+		semconv.GenAIResponseFinishReasons(reasons),
+		semconv.GenAIUsageInputTokens(resp.Usage.InputTokens),
+		semconv.GenAIUsageOutputTokens(resp.Usage.OutputTokens),
+		// The Messages API reports no total_tokens field; derive it so the
+		// span shape matches the other GenAI instrumentations.
+		semconv.GenAIUsageTotalTokens(resp.Usage.InputTokens+resp.Usage.OutputTokens),
```
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go]: **[Major] `total_tokens` and `input_tokens` are incorrect when prompt caching is used**

Anthropic's `input_tokens` field excludes cache tokens — cache reads and creations are reported separately in `cache_read_input_tokens` and `cache_creation_input_tokens`. This differs from OpenAI, where `prompt_tokens` already includes cached tokens, so the `input + output` formula works there but not here.

Example with caching:
```
input_tokens: 500, cache_creation_input_tokens: 1000, output_tokens: 200
Expected total = 500 + 1000 + 200 = 1700
Actual (code)   = 500 + 200             = 700  ← underreports by 1000
```

Similarly, `gen_ai.usage.input_tokens` should include cache tokens per the semconv definition ("number of tokens used in the prompt").

Suggested fix:
```go
totalInput := resp.Usage.InputTokens +
    resp.Usage.CacheReadInputTokens +
    resp.Usage.CacheCreationInputTokens

span.SetAttributes(
    semconv.GenAIUsageInputTokens(totalInput),
    semconv.GenAIUsageOutputTokens(resp.Usage.OutputTokens),
    semconv.GenAIUsageTotalTokens(totalInput + resp.Usage.OutputTokens),
)
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+			span.End()
+			return resp, nil
+		}
+
+		contentType := resp.Header.Get("Content-Type")
+		isStreaming := strings.HasPrefix(contentType, "text/event-stream")
+
+		if isStreaming {
+			span.SetAttributes(semconv.GenAIRequestIsStream(true))
+			resp.Body = newStreamingReader(resp.Body, span, start)
```
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go]: **[Major] Streaming requests emit incomplete spans missing all response/usage attributes**

The middleware creates a span and sets `gen_ai.request.is_stream=true` for streaming responses, but `streamingReader.finalize()` only calls `span.End()` without setting any `gen_ai.response.*` or `gen_ai.usage.*` attributes. This means production users of the streaming Messages API will export gen_ai spans that are missing token usage data.

Downstream dashboards/alerts that expect `gen_ai.usage.input_tokens` on every gen_ai span (as validated by `RequireGenAIClientSemconv`) will show data gaps for all streaming requests.

Since the PR title says "non-streaming", consider one of:
1. **Pass-through** (skip span creation for streaming) — treat it like `opUnknown` until accumulation is implemented:
```go
if isStreaming {
    span.End()
    return resp, nil
}
```
2. **Keep the span but mark it** — add an attribute like `gen_ai.response.incomplete=true` so consumers can filter them.

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		provider := getProviderName(req.URL.Host)
+		opName := operationName(op)
+
+		// Read a bounded copy for attribute parsing, but preserve the full body for the SDK.
+		var buf bytes.Buffer
+		tee := io.TeeReader(req.Body, &buf)
+		bodyBytes, err := io.ReadAll(io.LimitReader(tee, maxRequestBodySize))
+		if err != nil {
+			return next(req)
+		}
```
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go]: **[Minor] Request body is partially consumed when `ReadAll` returns an error**

If `io.ReadAll` fails, `req.Body` has already been partially drained by the TeeReader, but the body is not reassembled before calling `next(req)`. This passes a truncated request body to the SDK, which will send a malformed request to the API.

While the body is typically a `bytes.Reader` that won't error in practice, a defensive fix is straightforward:

```go
if err != nil {
    req.Body = struct {
        io.Reader
        io.Closer
    }{io.MultiReader(&buf, req.Body), req.Body}
    return next(req)
}
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	_ context.Context,
+	resp *http.Response,
+	span trace.Span,
+	_ time.Time,
+) {
+	defer span.End()
+
+	// Read a bounded preview for parsing, but reassemble the full body for callers.
+	var buf bytes.Buffer
+	tee := io.TeeReader(resp.Body, &buf)
```
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go:172]: **[Minor] Missing nil check on `resp.Body`**

If `resp.Body` is nil, `io.TeeReader(resp.Body, &buf)` will create a TeeReader with a nil source reader, causing a nil pointer dereference panic when `ReadAll` calls `Read`.

In practice, the Anthropic SDK's HTTP transport always returns a non-nil body for successful responses, but adding a guard is cheap:

```go
if resp.Body == nil {
    return
}
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/streaming.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	if r.done.CompareAndSwap(false, true) {
+		r.finalize()
+	}
+	if r.reader != nil {
+		return r.reader.Close()
+	}
+	return nil
+}
+
+func (r *streamingReader) finalize() {
```
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/streaming.go]: **[Minor] `streamingReader` edge cases are untested**

The `streamingReader` contains real logic (atomic CAS guard for idempotent `finalize`, span lifecycle management), but the only test coverage is `TestOtelMiddleware_Streaming` in `middleware_test.go`, which only verifies the happy path. The following edge cases from the comparable OpenAI `streaming_test.go` are missing:

- **Close-before-Read**: caller closes `resp.Body` without reading (common in error paths) — span must still end exactly once
- **Multiple Close**: defensive double-close must not double-end the span
- **Empty stream**: empty body reaching EOF — span must end exactly once
- **Body integrity**: the test discards the read body content (`_, err = io.ReadAll(resp.Body)`); consider asserting the content matches the original SSE data

Consider adding a dedicated `streaming_test.go` mirroring the OpenAI test suite.

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+func getProviderName(host string) string {
+	for keyword, provider := range providerMapping {
+		if strings.Contains(host, keyword) {
+			return provider
+		}
+	}
+	return "anthropic"
+}
```
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go:43]: **[Suggestion] `getProviderName` uses map iteration with non-deterministic order**

Go map iteration order is randomized. If a host matches multiple keywords (e.g., a hypothetical `anthropic.com.localhost`), the returned provider would be non-deterministic. While the current three keywords are effectively mutually exclusive for valid hosts, using a map for ordered matching is a code smell.

Consider using an ordered slice for deterministic priority:

```go
var providerEntries = []struct{ keyword, provider string }{
    {"anthropic.com", "anthropic"},
    {"localhost", "local"},
    {"127.0.0.1", "local"},
}

func getProviderName(host string) string {
    for _, e := range providerEntries {
        if strings.Contains(host, e.keyword) {
            return e.provider
        }
    }
    return "anthropic"
}
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go (hunk 1 above)
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go]: Anthropic's `input_tokens` field excludes cache tokens — cache reads and creations are reported separately in `cache_read_input_tokens` and `cache_creation_input_tokens`. This differs from OpenAI, where `prompt_tokens` already includes cached tokens, so the `input + output` formula works there but not here.

Example with caching:
```
input_tokens: 500, cache_creation_input_tokens: 1000, output_tokens: 200
Expected total = 500 + 1000 + 200 = 1700
Actual (code)   = 500 + 200             = 700  ← underreports by 1000
```

Similarly, `gen_ai.usage.input_tokens` should include cache tokens per the semconv definition ("number of tokens used in the prompt").

Suggested fix:
```go
totalInput := resp.Usage.InputTokens +
    resp.Usage.CacheReadInputTokens +
    resp.Usage.CacheCreationInputTokens

span.SetAttributes(
    semconv.GenAIUsageInputTokens(totalInput),
    semconv.GenAIUsageOutputTokens(resp.Usage.OutputTokens),
    semconv.GenAIUsageTotalTokens(totalInput + resp.Usage.OutputTokens),
)
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go (hunk 2 above)
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go]: The middleware creates a span and sets `gen_ai.request.is_stream=true` for streaming responses, but `streamingReader.finalize()` only calls `span.End()` without setting any `gen_ai.response.*` or `gen_ai.usage.*` attributes. This means production users of the streaming Messages API will export gen_ai spans that are missing token usage data.

Downstream dashboards/alerts that expect `gen_ai.usage.input_tokens` on every gen_ai span (as validated by `RequireGenAIClientSemconv`) will show data gaps for all streaming requests.

Since the PR title says "non-streaming", consider one of:
1. **Pass-through** (skip span creation for streaming) — treat it like `opUnknown` until accumulation is implemented:
```go
if isStreaming {
    span.End()
    return resp, nil
}
```
2. **Keep the span but mark it** — add an attribute like `gen_ai.response.incomplete=true` so consumers can filter them.

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go (hunk 3 above)
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go]: If `io.ReadAll` fails, `req.Body` has already been partially drained by the TeeReader, but the body is not reassembled before calling `next(req)`. This passes a truncated request body to the SDK, which will send a malformed request to the API.

While the body is typically a `bytes.Reader` that won't error in practice, a defensive fix is straightforward:

```go
if err != nil {
    req.Body = struct {
        io.Reader
        io.Closer
    }{io.MultiReader(&buf, req.Body), req.Body}
    return next(req)
}
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go (hunk 4 above)
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go:172]: If `resp.Body` is nil, `io.TeeReader(resp.Body, &buf)` will create a TeeReader with a nil source reader, causing a nil pointer dereference panic when `ReadAll` calls `Read`.

In practice, the Anthropic SDK's HTTP transport always returns a non-nil body for successful responses, but adding a guard is cheap:

```go
if resp.Body == nil {
    return
}
```

File: instrumentation/github.com/anthropics/anthropic-sdk-go/streaming.go (hunk 5 above)
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/streaming.go]: The `streamingReader` contains real logic (atomic CAS guard for idempotent `finalize`, span lifecycle management), but the only test coverage is `TestOtelMiddleware_Streaming` in `middleware_test.go`, which only verifies the happy path. The following edge cases from the comparable OpenAI `streaming_test.go` are missing:

- **Close-before-Read**: caller closes `resp.Body` without reading (common in error paths) — span must still end exactly once
- **Multiple Close**: defensive double-close must not double-end the span
- **Empty stream**: empty body reaching EOF — span must end exactly once
- **Body integrity**: the test discards the read body content (`_, err = io.ReadAll(resp.Body)`); consider asserting the content matches the original SSE data

Consider adding a dedicated `streaming_test.go` mirroring the OpenAI test suite.

File: instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go (hunk 6 above)
Comment (NameHaibinZhang) [inline instrumentation/github.com/anthropics/anthropic-sdk-go/middleware.go:43]: Go map iteration order is randomized. If a host matches multiple keywords (e.g., a hypothetical `anthropic.com.localhost`), the returned provider would be non-deterministic. While the current three keywords are effectively mutually exclusive for valid hosts, using a map for ordered matching is a code smell.

Consider using an ordered slice for deterministic priority:

```go
var providerEntries = []struct{ keyword, provider string }{
    {"anthropic.com", "anthropic"},
    {"localhost", "local"},
    {"127.0.0.1", "local"},
}

func getProviderName(host string) string {
    for _, e := range providerEntries {
        if strings.Contains(host, e.keyword) {
            return e.provider
        }
    }
    return "anthropic"
}
```

File: test/integration/anthropic_client_test.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+					"text": "Hello!",
+				},
+			},
+			"stop_reason":   "end_turn",
+			"stop_sequence": nil,
+			"usage": map[string]any{
+				"input_tokens":  10,
+				"output_tokens": 20,
+			},
+		}
```
Comment (NameHaibinZhang) [inline test/integration/anthropic_client_test.go:127]: The mock server returns `input_tokens: 10, output_tokens: 20` but omits `cache_read_input_tokens` and `cache_creation_input_tokens`. The `RequireGenAIClientSemconv` helper only validates standard attributes. The Anthropic-specific cache token attributes (`gen_ai.usage.cache_read.input_tokens`, `gen_ai.usage.cache_creation.input_tokens`) are unit-tested but not verified end-to-end through the instrumented binary.

Consider adding a second test case whose mock server returns non-zero cache values and asserting those attributes appear on the span.

File: tool/data/otelc-bundle.tgz
Comment (NameHaibinZhang) [inline tool/data/otelc-bundle.tgz:1]: this is not use

**Comment by pradhyum6144** [comment]
Thanks for the thorough review! All findings addressed in 782b79c. Module coverage is now 100% with -race clean.

On tool/data/otelc-bundle.tgz  could you clarify? The Verify Bundle check regenerates the archive via make package and fails if the committed file differs, which is why it's included. Happy to drop it if there's a different workflow for new instrumentations.

**Review by NameHaibinZhang — APPROVED** [review]

**Comment by NameHaibinZhang** [comment]
A conversation must be resolved before this pull request can be merged.

**Comment by pradhyum6144** [comment]
Thanks @NameHaibinZhang! All conversations resolved, CI is green, and it's approved ready to merge whenever you are.

**Review by NameHaibinZhang — APPROVED** [review]

**Comment by amazingakai** [comment]
@pradhyum6144 Please also add a corresponding entry to the OTel registry (see [registry guide](https://github.com/open-telemetry/opentelemetry-go-compile-instrumentation/blob/main/docs/instrument-guide.md#4-register-the-instrumentation)).

**Comment by pradhyum6144** [comment]
> @pradhyum6144 Please also add a corresponding entry to the OTel registry (see [registry guide](https://github.com/open-telemetry/opentelemetry-go-compile-instrumentation/blob/main/docs/instrument-guide.md#4-register-the-instrumentation)).

Thanks @amazingakai opened open-telemetry/opentelemetry.io#10909 with the registry entry. I also added Anthropic to the compile-time supported-libraries page, since OpenAI was listed there but the Anthropic row was missing
